### PR TITLE
Language Detection

### PR DESCRIPTION
**Pull request by mateusz-wosinski-ds**
### Description

Adds language detection component based on [langdetect](https://github.com/Mimino666/langdetect/tree/master/langdetect) library. It implements naive Bayesian filter and is able to recognize >50 languages. This module can be especially useful together with components that require selection of the language (e.g. text-2-speech or data-anonymizer)

### Next steps

Add integration with [fasttext](https://fasttext.cc/docs/en/language-identification.html) library, which can achieve even better results in the language detection task.

### Twitter handle

@deepsense_ai, @matt_wosinski

### CODE REVIEW

File: libs/experimental/langchain_experimental/language_detector/lang_detect.py [hunk 1]
```diff
@@ -0,0 +1,34 @@
...
+            threshold  # Minimum probability to consider a language as detected.
+        )
+
+    def _detect_single(self, text: str) -> str:
+        """Detects the most probable language of a single text."""
+        return langdetect.detect(text)
+
+    def _detect_many(self, text: str) -> List[Tuple[str, float]]:
+        """Detects all languages of a single text with a score bigger than threshold.
+        Returns them sorted on score in descending order.
```
Comment (eryk-dsai) [inline libs/experimental/langchain_experimental/language_detector/lang_detect.py]: Does the `detect_langs` returns the languages sorted by the score ? If not, there is no logic implemented for sorting

**Review by eryk-dsai — APPROVED** [review]

File: libs/experimental/langchain_experimental/language_detector/lang_detect.py (hunk 1 above)
Comment (mateusz-wosinski-ds) [reply libs/experimental/langchain_experimental/language_detector/lang_detect.py]: yes, it sorts them in descending order. I've added an assertion in test to make it crystal clear :raised_hands:

File: docs/extras/guides/privacy/presidio_data_anonymization/multi_language.ipynb [hunk 2]
```diff
@@ -190,21 +190,99 @@
...
     "print(anonymizer.anonymize(\"My name is John\"))"
    ]
   },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "## Usage with other tools\n",
+    "\n",
+    "### Language detection\n",
```
Comment (maks-operlejn-ds) [inline docs/extras/guides/privacy/presidio_data_anonymization/multi_language.ipynb:214]: If we have it here, remove "automatic language detection" from future works - basically, you can delete future works section now XD

Comment (mateusz-wosinski-ds) [reply docs/extras/guides/privacy/presidio_data_anonymization/multi_language.ipynb:214]: done :+1:

**Review by maks-operlejn-ds — APPROVED** [review]

File: libs/experimental/langchain_experimental/language_detector/lang_detect.py [hunk 3]
```diff
@@ -0,0 +1,34 @@
...
+
+try:
+    import langdetect
+except ImportError:
+    raise ImportError(
+        "Could not import langdetect, please install with " "`pip install langdetect`."
+    )
+
+
+class LangDetector(LanguageDetectorBase):
```
Comment (piotr-grodek-dsai) [inline libs/experimental/langchain_experimental/language_detector/lang_detect.py]: I saw in notebooks that everyone would need to reimplement the same snippet:
```
def lang_transform_func(inputs: dict) -> dict:
    text = inputs["text"]
    language = lang_detector.detect_single_language(text)
    return {"text": text, "language": language}

lang_transform_chain = TransformChain(input_variables=["text"], output_variables=["text", "language"], transform=lang_transform_func)
```
At least it should be already defined here tbh to provide a value for me, as dev. I don't need/want to reimplement this snippet as it should work out of box smoothly for other elements.
@maks-operlejn-ds I don't remember correctly, but was there a review in your MR about not providing such functionalities out of box?

File: libs/experimental/langchain_experimental/language_detector/lang_detect.py [hunk 4]
```diff
@@ -0,0 +1,34 @@
+from typing import List, Tuple
+
+from langchain_experimental.language_detector.base import LanguageDetectorBase
+
+try:
+    import langdetect
```
Comment (piotr-grodek-dsai) [inline libs/experimental/langchain_experimental/language_detector/lang_detect.py]: I think people from langchain team want to have failure on use, but no error on import so it does not fail in tests when modules are importer.

File: libs/experimental/langchain_experimental/language_detector/lang_detect.py [hunk 5]
```diff
@@ -0,0 +1,34 @@
...
+class LangDetector(LanguageDetectorBase):
+    """
+    Language detector based on langdetect package.
+    Warning: Results are not deterministic!
+    It supports 55 languages out of the box.
+    """
+
+    def __init__(self, threshold: float = 0.1):
+        self.threshold = (
+            threshold  # Minimum probability to consider a language as detected.
```
Comment (piotr-grodek-dsai) [inline libs/experimental/langchain_experimental/language_detector/lang_detect.py]: I don't see much value in threshold here, what if I want to have a streamlit app with slider for it? it is ugly to re-create object to change threshold value. Also the default I expect to see is zero - it should return all detected languages. Also, function at line 25 does not use threshold nor return it - it already is confusing.

**Review by piotr-grodek-dsai — COMMENTED** [review]
Left some comments.

**Pull request closed without merging**